### ISSUE TITLE
Unable to retrieve annotations for Kasthuri CELL (2014) dataset

### ISSUE THREAD

**Issue by neurohn** [openconnectome/ndio#194]
| ndio version | os | python |
| --- | --- | --- |
| 1.1.0 | Ubuntu 14.04 LTS | 2.7.6 |

I'm trying to download annotations for Kasthuri CELL (2014) dataset by following [this Ipython notebook](https://github.com/neurodata/ndio-demos/blob/master/Datasets/Kasthuri%20et%20al%20%282015%29%20%E2%80%94%20Mouse%20S1.ipynb) recently uploaded on ndio-demos repo.

This code works just fine

```
query = {
    'token': human_annotation_token,
    'channel': 'synapses',
    'x_start': 1000,
    'x_stop': 1300,
    'y_start': 2000,
    'y_stop': 2400,
    'z_index': 1100,
    'resolution': 1
}

slic = nd.get_image(**query)
```

However this one doesn't

```
query = {
    'token': human_annotation_token,
    'channel': 'synapses',
    'x_start': 0,
    'x_stop': 10752,
    'y_start': 0,
    'y_stop': 13312,
    'z_index': 1100,
    'resolution': 1
}

slic = nd.get_image(**query)
```

As you might observe I'm trying to download the full tile at a z-slice at highest resolution. I get the dimension limits using the block of code:

```
data = nd.get_proj_info(token)
imagesSizes = data['dataset']['imagesize']
offset = data['dataset']['offset']

[x_start, x_stop, y_start, y_stop, z_start, z_stop] = [offset[n][0],
    imagesSizes[n][0], offset[n][1],imagesSizes[n][1], offset[n][2],
    imagesSizes[n][2]]
```

This is the error I get

```
IOError: Bad server response for http://openconnecto.me/ocp/ca/kasthuri2015_ramon_v4/synapses/hdf5/1/0,10752/0,13312/1,2/: 502: <!DOCTYPE html>
<html>
<head>
<title>Error</title>
<style>
body {
width: 35em;
margin: 0 auto;
font-family: Tahoma, Verdana, Arial, sans-serif;
}
</style>
</head>
<body>
<h1>An error occurred.</h1>
<p>Sorry, the page you are looking for is currently unavailable.<br/>
Please try again later.</p>
<p>If you are the system administrator of this resource then you should check
the <a href="http://nginx.org/r/error_log">error log</a> for details.</p>
<p><em>Faithfully yours, nginx.</em></p>
</body>
</html>
```

Am I missing something obvious here? I'd be grateful if someone can look into this.
@willgray @j6k4m8

**Comment by wrgr** [comment]
Hi Rohan!

Thanks for the bug report - that was very thorough and helpful.

@j6k4m8 - can you take a look at the query itself?  It's big, but should
chunk.

I'll take a look at the content itself (e.g., the data you are pulling has
no synapses for most of the volume, but we can help!)

Bear with us - it's graduation week at JHU and we're wrapping up some short
suspense papers.

**Comment by j6k4m8** [comment]
Interesting — this error is displayed when there is a _server-side_ error, though it also looks like the ndio request is a bit outrageous. (It's under the memory cap, but... still pretty high.) 

As a quick fix, I'd try changing your 

```
nd = neurodata()
```

to 

```
nd = neurodata(chunk_threshold=(1E9 / 8))
```

which should help section off downloads into 1/8-GB-sized chunks (all invisible to you, but helps with some corner-cases like this one). **Note that this slice is totally empty**, to the best of my knowledge.

Setting `chunk_threshold` to something really silly, like `1000`, is a great way to sanity-check downloads, fwiw.

**Comment by neurohn** [comment]
Sure no problem @willgray 
Thanks for the tip @j6k4m8 . But this hack still gives me the same error. It doesn't seem to me that the problem is the size of the slice because the same slice from 'kasthuri11' dataset is 143.1MB; so not too large right? Let me know your thoughts? Thanks again for attending to this issue.

**Comment by wrgr** [comment]
should be able to get an example script for synapses using ndio by early next week latest.

**Comment by neurohn** [comment]
Was anybody able to figure this out? I still get the same bug on the latest version of ndio.

**Comment by wrgr** [comment]
This fell off our stack I think.  Thanks for the bump.  @j6k4m8 - this is related to https://github.com/neurodata/ndio/issues/192

**Comment by Aeusman** [comment]
Interesting that: 
http://openconnecto.me/ocp/ca/kasthuri2015_ramon_v4/synapses/xy/0/0,10000/0,12000/2/ 
works but:
http://openconnecto.me/ocp/ca/kasthuri2015_ramon_v4/synapses/xy/0/0,10000/0,14000/2/
does not. I'm wondering what the cutoff for that is?

Reason being that right now the size is being calculated as 0.1 gigs, which is below the threshold and so neurodata module doesn't chunk. But the backend apparently sees that size as too large (or some other error) and throws some error.

**Comment by j6k4m8** [comment]
@Aeusman — that's a _super_ good catch. I imagine downloading in tiles and then stitching together would be a good interim shim, but a real fix would require figuring out why the backend selectively returns or does not return a slice.

**Comment by wrgr** [comment]
@kunallillaney - do you know why this fails?  

http://openconnecto.me/ocp/ca/kasthuri2015_ramon_v4/synapses/xy/0/0,10000/0,14000/2/

I wonder if this is actually an ndstore thing rather than an ndio thing - or maybe we are misunderstanding an implementation detail in ndio.  This restful call is ostensibly 560MB at 4 bytes/pixel, but should compress nearly to zero (this slice is blank).  Also happens when using get_cutout or get_slice interfaces.

**Comment by Aeusman** [comment]
The reason is that the data size is too large to fit in the memory allocated for the RESTful call on the backend which is why it throws an error, meaning we might need to either reduce the threshold further (less preferred by kunal) or just always chunk the data like with larger pulls (more preferred by kunal). Also regardless of what we pick we need to add the datatype/size to the threshhold, e.g. 10000x10000x32=0.4 GB instead of just 10000x10000=0.1 GB (which is how its currently calculated)

**Comment by wrgr** [comment]
i thought we were good up to 2GB?  We should add that - if it's not there, let's add datatype size into computation as an issue (or PR!)

Chunking all the time is much slower given our currently available interfaces and overhead.  Will slow down dataset scraping by a factor of 5-10 (?) based on old tests that we can rerun as needed.

**Comment by kunallillaney** [comment]
@willgray This is where there is a fundamental misunderstanding of ndstore. Yes you are good to go till 2GB but the way you compute that size is incorrect. You do a simple calculation of 10,000x14,000x4 = 560 MB wherein ndstore still has to do cuboid cutouts in multiple of 128x128x16 which translates into 12,800x14,080x16 = 8.5GB. This sort of pertains to the larger discussion we had yesterday.
With respect to chunking, did you guys test chunking with multiple info calls, then yes it will be slow because of the inefficient access pattern of info for every single chunk. But if chunking is done in the correct sense then it should be faster then accessing a single large blob. (You still pay some time for de-serialization and rearrangement on client side but that is offset by parallelism).

**Comment by wrgr** [comment]
good point on the info_calls - I need to PR that in, today or over the weekend.

You mean `10112 x 14080 x 16` - right?  I see - we definitely realized that you had to cutout on the backend, but didn't realize that you limited the bigger cutout to the 2GB figure.  That's helpful.  Also trickier to compute.  Maybe we should do something like ceiling to nearest multiple of cube size.  Won't be quite right if we aren't cube aligned, but is a closer approximation.  

This is an ndio problem as things stand.  Definitely not obvious to the end user, so someone (maybe ndio) should explain this.  

Take away here is that we need to somehow put the max size into a form that users can grok.  
_Pulling down a single slice of data is a thing that neurodata/ndstore should support trivially._

**Comment by kunallillaney** [comment]
ndstore does support pulling down single slice of data trivially. That is the xy/yz/xz/tile interface we have. But it is inefficient because we store our data as cuboids. The only way for anyone(including the back-end itself) to read data from the database is in multiples of 128x128x16. You can still pull a single slice but it will be slow when compared to pulling 16 of them at the same time because it is 16 versus single I/O you are doing.
We cannot make ndstore efficient for both volume and single slice cutouts because the system can only be designed in one way to avoid storing data multiple times. Having said that we allow both of these cutouts for the ease of the user. If you are going to be a power user and want to speed up your workflow then you have to design your workflow based on the system's design, the system cannot be designed to accommodate everyone's workflow.

**Comment by kunallillaney** [comment]
Also, it is not hard to compute what I just did for you. All you need is cuboid sizes for that resolution and data offsets(which are available via info call). If @j6k4m8 wants to write this correctly, I can tell him the math behind it.

**Comment by wrgr** [comment]
@kunallillaney:  dude - yes, we know this.  :)  The non-intuitive part is when slices are too large, because of the cuboids.  It's not intuitive for all of our end-users.  No action required from you!

@Aeusman is going to take a shot at a PR.

**Comment by wrgr** [comment]
Initializing ndio like this:   `nd = neurodata(chunk_threshold=(1E9 / 8))`  did indeed work for me, but was incredibly slow.  We're working on a PR to try to reduce the queries behind the scenes.  :)  And will report back.  

@neurohn - in the meantime, what are you actually trying to get in terms of data?  Bug is noted and will be fixed!  But worth noting that the data you are trying to pull is blank, so perhaps not so useful in the end.  Can I point you to a more useful sample to get you moving in parallel?

**Comment by Aeusman** [comment]
Fixed and merged in #216
